README.md
# F5 Systems Logs

### Overview
This repository is a parser for F5 Syslogs. 

Data Log
```sh
  = LogSslRequest SslRequest
  | LogSslAccess SslAccess
  | LogAsm Asm
  | LogAsmKeyValue (Chunks Attribute)
```

Data Attributes
```sh
  = Action {-# UNPACK #-} !Bytes
  | AttackType {-# UNPACK #-} !Bytes
  | DestinationIp {-# UNPACK #-} !IPv4 -- ^ The F5 IP address, not the server IP address.
  | DestinationPort {-# UNPACK #-} !Word16
  | GeoLocation {-# UNPACK #-} !Bytes -- ^ Two-letter country code
  | Headers !(Chunks Header)
  | IpClient {-# UNPACK #-} !IPv4 -- ^ IP address of the client, @ip_client@.
  | ManagementIpAddress {-# UNPACK #-} !IPv4 -- ^ IP address of F5.
  | Protocol {-# UNPACK #-} !Bytes
  | RequestBody {-# UNPACK #-} !Bytes
  | RequestMethod {-# UNPACK #-} !Bytes
  | RequestStatus {-# UNPACK #-} !Bytes
  | RequestTarget {-# UNPACK #-} !Bytes
  | ResponseCode {-# UNPACK #-} !Word64
  | Scheme {-# UNPACK #-} !Bytes
  | Severity {-# UNPACK #-} !Bytes
  | SourcePort {-# UNPACK #-} !Word16
  | SupportId {-# UNPACK #-} !Word64
  | Uri {-# UNPACK #-} !Bytes
  | Username {-# UNPACK #-} !Bytes
```

CHANGELOG.md
# Revision history for f5-syslog

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/Main.hs
{-# language DuplicateRecordFields #-}
{-# language MultiWayIf #-}
{-# language NamedFieldPuns #-}
{-# language TypeApplications #-}

import F5.Syslog (Log(..),SslAccess(..),SslRequest(..),Asm(..),decode)

import Control.Exception (throwIO)
import Data.Primitive (ByteArray)
import Data.Word (Word8)
import Data.Char (ord)
import Data.Bytes.Types (Bytes(Bytes))

import qualified Data.Primitive as PM
import qualified GHC.Exts as Exts
import qualified Net.IPv4 as IPv4
import qualified Sample as S

main :: IO ()
main = do
  putStrLn "Start"
  putStrLn "testSslAccess1"
  testSslAccess1
  putStrLn "testSslRequest1"
  testSslRequest1
  putStrLn "testSslAsm1"
  testAsm1
  putStrLn "End"

testSslAccess1 :: IO ()
testSslAccess1 = case decode S.ssl_access_1 of
  Left err -> throwIO err
  Right (LogSslAccess SslAccess{path,user,responseCode,responseBytes,client,host}) ->
    if | path /= bytes "/iControl/iControlPortal.cgi" -> fail "bad path"
       | user /= bytes "" -> fail "bad user"
       | responseCode /= 200 -> fail "bad response code"
       | responseBytes /= 626 -> fail "bad response bytes"
       | client /= IPv4.fromOctets 127 0 0 1 -> fail "bad client"
       | host /= bytes "SAMPLE-HOST" -> fail "bad host"
       | otherwise -> pure ()
  Right _ -> fail "wrong log type"

testSslRequest1 :: IO ()
testSslRequest1 = case decode S.ssl_request_1 of
  Left err -> throwIO err
  Right (LogSslRequest SslRequest{path,protocol,cipherSuite}) ->
    if | path /= bytes "/iControl/iControlPortal.cgi" -> fail "bad path"
       | cipherSuite /= bytes "AES256-SHA" -> fail "bad cipher suite"
       | protocol /= bytes "TLSv1" -> fail "bad ssl protocol"
       | otherwise -> pure ()
  Right _ -> fail "wrong log type"

testAsm1 :: IO ()
testAsm1 = case decode S.asm_1 of
  Left err -> throwIO err
  Right (LogAsm Asm{destinationPort}) ->
    if | destinationPort /= 443 -> fail "bad destination port"
       | otherwise -> pure ()
  Right _ -> fail "wrong log type"

bytes :: String -> Bytes
bytes s = let b = pack s in Bytes b 0 (PM.sizeofByteArray b)

pack :: String -> ByteArray
pack = Exts.fromList . map (fromIntegral @Int @Word8 . ord)



common/Sample.hs
{-# language TypeApplications #-}

module Sample
  ( ssl_access_1
  , ssl_request_1
  , asm_1
  ) where

import Data.Bytes (Bytes)
import Data.Word (Word8)
import Data.Char (ord)
import qualified Data.Bytes as Bytes
import qualified GHC.Exts as Exts

-- Sample Logs. If you add a sample log to this file, please
-- replace all information in the log that could possibly be
-- meaningful. At a bare minimum, this means:
--
-- * Replace any IP addresses with non-routable addresses 
--   from the TEST-NET-1 block (192.0.2.0/24).
-- * Replace any domain names with the reserved domain
--   name example.com.
-- * Replace any hostnames with something like MY-HOST
--   or NY-APP or SAMPLE-HOST.

pack :: String -> Bytes
pack = Bytes.fromByteArray . Exts.fromList . map (fromIntegral @Int @Word8 . ord)

ssl_access_1 :: Bytes
ssl_access_1 = pack $ concat
  [ "Dec 6 10:02:54 SAMPLE-HOST info logger: "
  , "[ssl_acc] 127.0.0.1 - - [06/Dec/2019:10:02:54 -0600] "
  , "\"/iControl/iControlPortal.cgi\" 200 626"
  ]

ssl_request_1 :: Bytes
ssl_request_1 = pack $ concat
  [ "Nov 13 10:02:54 SAMPLE-HOST info logger: "
  , "[ssl_req][06/Dec/2019:10:02:54 -0600] 127.0.0.1 TLSv1 "
  , "AES256-SHA \"/iControl/iControlPortal.cgi\" 626"
  ]

asm_1 :: Bytes
asm_1 = pack $ concat
  [ "Dec 6 13:14:50 f5.example.com ASM:\"Information Leakage\","
  , "\"2019-12-06 13:14:49\",\"192.0.2.56\",\"443\",\"US\","
  , "\"/path/to/file\",\"N/A\",\"192.0.2.51\",\"192.0.2.117%0\","
  , "\"192.0.2.57\",\"GET\",\"2016-12-31 09:35:53\",\"/path/to/another\","
  , "\"HTTPS\",\"\",\"GET /foo/bar/baz HTTP/1.1\\r\\nHost: example.com"
  , "\\r\\nConnection: keep-alive\\r\\nAccept: text/html,"
  , "application/xhtml+xml,application/xml;q=0.9,*/*;"
  , "q=0.8\\r\\nUser-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 12_3_1 "
  , "like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) "
  , "Version/12.1.1 Mobile/15E148 Safari/604.1\\r\\nAccept-Language: "
  , "en-us\\r\\nReferer: https://www.example.com/foo/bar/fizz.seam?"
  , "cid=24313\\r\\nAccept-Encoding: br, gzip, deflate\\r\\n\\r\\n\","
  , "\"blocked\",\"Only illegal requests are logged\",\"403\",\"0\","
  , "\"2a0276a17b6b7d70\",\"Informational\",\"\",\"\",\"4131\",\"\","
  , "\"15267537619763049516\",\"f5.example.com\",\"/foo/bar/baz\","
  , "\"N/A\",\"\",\"Illegal HTTP status in response\",\"N/A\",\"N/A\""
  ]

-- Incomplete Reference Parser
-- 
-- F5-ASM[device_type=F5]: {!alpha} {!digit} {!digit}:{!digit}:{!digit} |
-- {host.name} ASM: "{threat_name}","{!digit}-{!digit}-{!digit} {!digit}:|
-- {!digit}:{!digit}","{destination.ip}","{destination.port}","{!alpha}",|
-- "/Common/HTTP?{S}-{http.domain}","{!drop}","{source.ip}","{!drop}",|
-- "{!drop}","{http.method}","{!digit}-{!digit}-{!digit} {!digit}:{!digit}|
-- :{!digit}","/Common/HTTP?{S}-{!drop}","{http.protocol}","{!drop}",|
-- "{!alpha} {http.path} HTTP?{S}/{http.version}\r\n{!drop}" ," {action}",|
-- "{message}","{http.status_code}","{!drop}","{!drop}","{severity}",|
-- "{!drop}","{!drop}","{source.port}","{!drop}","{!digit}","{!drop}",|
-- "{!drop}","?{N/A}?{{user}}","{!drop}","{http.message}","{threat_name}",|
-- "{!drop}"#{!digit}


src/F5/Syslog.hs
{-# language BangPatterns #-}
{-# language DerivingStrategies #-}
{-# language DeriveAnyClass #-}
{-# language DuplicateRecordFields #-}
{-# language LambdaCase #-}
{-# language NamedFieldPuns #-}

module F5.Syslog
  ( Log(..)
  , Asm(..)
  , SslRequest(..)
  , SslAccess(..)
  , decode
  ) where

import Control.Exception (Exception)
import Data.Builder.ST (Builder)
import Data.Bytes.Types (Bytes(Bytes))
import Data.Bytes.Parser (Parser,Result(Success,Failure),Slice(Slice))
import Data.Chunks (Chunks)
import Data.Word (Word16,Word64)
import Net.Types (IPv4)

import qualified Net.IPv4 as IPv4
import qualified Data.Builder.ST as Builder
import qualified Data.Bytes.Parser as P
import qualified Data.Bytes.Parser.Latin as Latin
import qualified Data.Bytes.Parser.Unsafe as Unsafe

data Log
  = LogSslRequest SslRequest
  | LogSslAccess SslAccess
  | LogAsm Asm

data Asm = Asm
  { destinationIp :: {-# UNPACK #-} !IPv4
  , destinationPort :: {-# UNPACK #-} !Word16
  , sourceIp :: {-# UNPACK #-} !IPv4
  , requestMethod :: {-# UNPACK #-} !Bytes
  , scheme :: {-# UNPACK #-} !Bytes
    -- ^ Examples: @HTTP@, @HTTPS@
  , headers :: !(Chunks Header)
  , action :: {-# UNPACK #-} !Bytes
  , responseCode :: {-# UNPACK #-} !Word64
  , requestTarget :: {-# UNPACK #-} !Bytes
  }

-- | The fields of a log of the form:
--
-- > info logger: [ssl_req][06/Dec/2019:10:02:54 -0600] 127.0.0.1 TLSv1 AES256-SHA "/iControl/iControlPortal.cgi" 626
data SslRequest = SslRequest
  { protocol :: {-# UNPACK #-} !Bytes
    -- ^ Examples: @TLSv1@
  , cipherSuite :: {-# UNPACK #-} !Bytes
    -- ^ Examples: @AES256-SHA@
  , path :: {-# UNPACK #-} !Bytes
    -- ^ Examples: @/iControl/iControlPortal.cgi@
  }

-- | The fields of a log of the form:
--
-- > info logger: [ssl_acc] 127.0.0.1 - - [06/Dec/2019:10:02:54 -0600] "/iControl/iControlPortal.cgi" 200 626
data SslAccess = SslAccess
  { path :: {-# UNPACK #-} !Bytes
    -- ^ Examples: @/iControl/iControlPortal.cgi@
  , user :: {-# UNPACK #-} !Bytes
    -- ^ Examples: @admin@, a hyphen becomes the empty string
  , responseCode :: {-# UNPACK #-} !Word64
    -- ^ Examples: 200, 404
  , responseBytes :: {-# UNPACK #-} !Word64
    -- ^ Examples: 672
  , client :: {-# UNPACK #-} !IPv4
    -- ^ Examples: 127.0.0.1
  , host :: {-# UNPACK #-} !Bytes
    -- ^ Examples: @lb.example.com@, @F5-Appliance-A@
  }

data Header = Header
  { name :: {-# UNPACK #-} !Bytes
  , value :: {-# UNPACK #-} !Bytes
  }

data Error
  = AsmDatetime
  | ClientEndOfInput
  | CountryEndOfInput
  | EncounteredLeftovers
  | HostEndOfInput
  | HttpPath
  | HttpTimestamp
  | LeadingDatetimeDay
  | LeadingDatetimeHour
  | LeadingDatetimeMinute
  | LeadingDatetimeMonth
  | LeadingDatetimeSecond
  | MalformedAction
  | MalformedApplianceIp
  | MalformedClientIdentity
  | MalformedClientIp
  | MalformedDestinationIp
  | MalformedDestinationPort
  | MalformedHeaderName
  | MalformedHeaderValue
  | MalformedHttpPath
  | MalformedHttpRequestMethod
  | MalformedHttpVersion
  | MalformedMethod
  | MalformedRequestTarget
  | MalformedResponseCode
  | MalformedScheme
  | MalformedSecondarySourceIp
  | MalformedSecondaryTimestamp
  | MalformedSourceIp
  | MissingIdentifier
  | MissingNewlineAfterHeader
  | MissingNewlineAfterLastHeader
  | PathDomainEndOfInput
  | ResponseBytes
  | ResponseCode
  | SecondaryPathDomainEndOfInput
  | SslCipherSuite
  | SslProtocol
  | ThreatEndOfInput
  | UnknownFieldA
  | UnknownFieldB
  | UnknownFieldC
  | UnknownFieldD
  | UnknownFieldE
  | UnknownFieldF
  | UnknownFieldG
  | UnknownFieldH
  | UnknownFieldI
  | UnknownFieldJ
  | UnknownFieldK
  | UnknownFieldL
  | UnknownFieldM
  | UnknownFieldN
  | UnknownFieldO
  | UnknownFieldP
  | UnknownFieldQ
  | UnrecognizedIdentifier
  deriving stock (Show)
  deriving anyclass (Exception)

decode :: Bytes -> Either Error Log
decode b = case P.parseBytes parser b of
  -- TODO: switch to parseBytesEither once it is released
  P.Failure e -> Left e
  P.Success (Slice _ _ r) -> Right r

parser :: Parser Error s Log
parser = do
  skipInitialDate
  !host <- P.takeTrailedBy HostEndOfInput 0x20 -- space
  Latin.any MissingIdentifier >>= \case
    'i' -> do
      Latin.char3 UnrecognizedIdentifier 'n' 'f' 'o'
      Latin.char UnrecognizedIdentifier ' '
      Latin.char7 UnrecognizedIdentifier 'l' 'o' 'g' 'g' 'e' 'r' ':'
      Latin.char UnrecognizedIdentifier ' '
      Latin.char5 UnrecognizedIdentifier '[' 's' 's' 'l' '_'
      Latin.any MissingIdentifier >>= \case
        'r' -> do
          Latin.char3 UnrecognizedIdentifier 'e' 'q' ']'
          parserSslRequest host
        'a' -> do
          Latin.char4 UnrecognizedIdentifier 'c' 'c' ']' ' '
          parserSslAccess host
        _ -> P.fail UnrecognizedIdentifier
    'A' -> do
      Latin.char4 UnrecognizedIdentifier 'S' 'M' ':' '"'
      parserAsm host
    _ -> P.fail UnrecognizedIdentifier

parserAsm :: Bytes -> Parser Error s Log
parserAsm !host = do
  P.takeTrailedBy ThreatEndOfInput 0x22 -- double quote
  Latin.char2 ThreatEndOfInput ',' '"'
  -- Skip the datetime
  Latin.skipTrailedBy AsmDatetime '"'
  Latin.char2 AsmDatetime ',' '"'
  destinationIp <- IPv4.parserUtf8Bytes MalformedDestinationIp
  quoteCommaQuote MalformedDestinationIp
  destinationPort <- Latin.decWord16 MalformedDestinationPort
  quoteCommaQuote MalformedDestinationPort
  skipPlainField CountryEndOfInput
  skipPlainField PathDomainEndOfInput
  skipPlainField UnknownFieldA
  sourceIp <- IPv4.parserUtf8Bytes MalformedSourceIp
  quoteCommaQuote MalformedSourceIp
  skipPlainField MalformedSecondarySourceIp -- has percent zero at end, weird
  skipPlainField MalformedApplianceIp
  requestMethod <- takePlainField MalformedMethod
  skipAsmDatetime
  quoteCommaQuote MalformedSecondaryTimestamp
  skipPlainField SecondaryPathDomainEndOfInput
  scheme <- takePlainField MalformedScheme
  skipPlainField UnknownFieldB
  -- Now the big one, the HTTP request and headers.
  -- We skip the request method since it showed up earlier
  Latin.skipTrailedBy MalformedHttpRequestMethod ' '
  -- We skip the path since it shows up later
  Latin.skipTrailedBy MalformedHttpPath ' '
  -- Skip the HTTP version
  Latin.char5 MalformedHttpVersion 'H' 'T' 'T' 'P' '/'
  match isDigit MalformedHttpVersion
  Latin.char MalformedHttpVersion '.'
  match isDigit MalformedHttpVersion
  Latin.char4 MalformedHttpVersion '\\' 'r' '\\' 'n'
  -- Parse all the headers
  headers <- allHeaders =<< P.effect Builder.new
  quoteCommaQuote MalformedAction
  action <- takePlainField MalformedAction
  skipPlainField UnknownFieldC
  responseCode <- Latin.decWord64 MalformedResponseCode
  quoteCommaQuote MalformedAction
  skipPlainField UnknownFieldD
  skipPlainField UnknownFieldE
  skipPlainField UnknownFieldF
  skipPlainField UnknownFieldG
  skipPlainField UnknownFieldH
  skipPlainField UnknownFieldI
  skipPlainField UnknownFieldJ
  skipPlainField UnknownFieldK
  skipPlainField UnknownFieldL -- F5 hostname
  requestTarget <- takePlainField MalformedRequestTarget
  skipPlainField UnknownFieldM
  skipPlainField UnknownFieldN
  skipPlainField UnknownFieldO
  skipPlainField UnknownFieldP
  Latin.skipTrailedBy UnknownFieldQ '"'
  pure $ LogAsm $ Asm
    { destinationIp , destinationPort , sourceIp
    , requestMethod , scheme , headers , action
    , responseCode , requestTarget
    }

allHeaders :: Builder s Header -> Parser Error s (Chunks Header)
allHeaders !b0 = Latin.trySatisfy (== '\\') >>= \case
  True -> do
    Latin.char3 MissingNewlineAfterLastHeader 'r' '\\' 'n'
    P.effect (Builder.freeze b0)
  False -> do
    !hdr <- oneHeader
    b1 <- P.effect (Builder.push hdr b0)
    allHeaders b1

oneHeader :: Parser Error s Header
oneHeader = do
  name <- P.takeTrailedBy MalformedHeaderName 0x3A -- colon
  Latin.skipChar ' '
  value <- P.takeTrailedBy MalformedHeaderValue 0x5C -- backslash
  Latin.char3 MissingNewlineAfterHeader 'r' '\\' 'n'
  pure Header{name,value}

skipPlainField :: e -> Parser e s ()
skipPlainField e = do
  Latin.skipTrailedBy e '"'
  Latin.char2 e ',' '"'

takePlainField :: e -> Parser e s Bytes
takePlainField e = do
  r <- P.takeTrailedBy e 0x22 -- double quote
  Latin.char2 e ',' '"'
  pure r

quoteCommaQuote :: e -> Parser e s ()
quoteCommaQuote e =
  Latin.char3 e '"' ',' '"'

parserSslAccess :: Bytes -> Parser Error s Log
parserSslAccess !host = do
  client <- IPv4.parserUtf8Bytes MalformedClientIp
  Latin.char MalformedClientIp ' '
  -- Client identity from identd is never used.
  Latin.char2 MalformedClientIdentity '-' ' '
  user <- Latin.trySatisfy (== '-') >>= \case
    True -> do
      Latin.char ClientEndOfInput ' '
      emptyBytes
    False -> P.takeTrailedBy ClientEndOfInput 0x20
  Latin.char HttpTimestamp '['
  Latin.skipTrailedBy HttpTimestamp ']'
  Latin.char HttpTimestamp ' '
  Latin.char HttpPath '"'
  path <- P.takeTrailedBy HttpPath 0x22 -- double quote
  Latin.char HttpPath ' '
  responseCode <- Latin.decWord64 ResponseCode
  Latin.char ResponseCode ' '
  responseBytes <- Latin.decWord64 ResponseBytes
  P.endOfInput EncounteredLeftovers
  pure (LogSslAccess (SslAccess {path,user,responseCode,responseBytes,client,host}))

parserSslRequest :: Bytes -> Parser Error s Log
parserSslRequest !host = do
  -- Skip the datetime information
  Latin.char HttpTimestamp '['
  Latin.skipTrailedBy HttpTimestamp ']'
  Latin.char HttpTimestamp ' '
  _ <- IPv4.parserUtf8Bytes MalformedClientIp
  Latin.char MalformedClientIp ' '
  protocol <- P.takeTrailedBy SslProtocol 0x20 -- space
  cipherSuite <- P.takeTrailedBy SslCipherSuite 0x20 -- space
  Latin.char HttpPath '"'
  path <- P.takeTrailedBy HttpPath 0x22 -- double quote
  Latin.char HttpPath ' '
  _ <- Latin.decWord64 ResponseBytes
  P.endOfInput EncounteredLeftovers
  pure (LogSslRequest (SslRequest {protocol,cipherSuite,path}))

emptyBytes :: Parser e s Bytes
emptyBytes = do
  arr <- Unsafe.expose
  pure (Bytes arr 0 0)

-- The initial datetime is formatted like this: Dec 6 10:04:50.
-- It is always missing the year. This consumes a trailing space.
skipInitialDate :: Parser Error s ()
skipInitialDate = do
  match isUpper LeadingDatetimeMonth
  match isLower LeadingDatetimeMonth
  match isLower LeadingDatetimeMonth
  Latin.char LeadingDatetimeMonth ' '
  Latin.skipDigits1 LeadingDatetimeDay
  Latin.char LeadingDatetimeDay ' '
  match isDigit LeadingDatetimeHour
  match isDigit LeadingDatetimeHour
  Latin.char LeadingDatetimeHour ':'
  match isDigit LeadingDatetimeMinute
  match isDigit LeadingDatetimeMinute
  Latin.char LeadingDatetimeMinute ':'
  match isDigit LeadingDatetimeSecond
  match isDigit LeadingDatetimeSecond
  Latin.char LeadingDatetimeSecond ' '

-- The standard YYYY-MM-DD HH:mm:SS format.
skipAsmDatetime :: Parser Error s ()
skipAsmDatetime = do
  match isDigit MalformedSecondaryTimestamp
  match isDigit MalformedSecondaryTimestamp
  match isDigit MalformedSecondaryTimestamp
  match isDigit MalformedSecondaryTimestamp
  Latin.char MalformedSecondaryTimestamp '-'
  match isDigit MalformedSecondaryTimestamp
  match isDigit MalformedSecondaryTimestamp
  Latin.char MalformedSecondaryTimestamp '-'
  match isDigit MalformedSecondaryTimestamp
  match isDigit MalformedSecondaryTimestamp
  Latin.char MalformedSecondaryTimestamp ' '
  match isDigit MalformedSecondaryTimestamp
  match isDigit MalformedSecondaryTimestamp
  Latin.char MalformedSecondaryTimestamp ':'
  match isDigit MalformedSecondaryTimestamp
  match isDigit MalformedSecondaryTimestamp
  Latin.char MalformedSecondaryTimestamp ':'
  match isDigit MalformedSecondaryTimestamp
  match isDigit MalformedSecondaryTimestamp

isUpper :: Char -> Bool
isUpper c = c >= 'A' && c <= 'Z'

isLower :: Char -> Bool
isLower c = c >= 'a' && c <= 'z'

isDigit :: Char -> Bool
isDigit c = c >= '0' && c <= '9'

match :: (Char -> Bool) -> Error -> Parser Error s ()
{-# inline match #-}
match p e = do
  c <- Latin.any e
  case p c of
    True -> pure ()
    False -> P.fail e
